spark/src/main/scala/geotrellis/spark/clip/ClipToGrid.scala

```diff
@@ -147,59 +150,28 @@ object ClipToGrid {
       clipFeature(extent, feature, preds).map(k -> _)
     }
 
-    val iterator: Iterator[(SpatialKey, Feature[Geometry, D])] =
+    val iterator: Iterator[(SpatialKey, Feature[Geometry, D])] = {
       feature.geom match {
-        case p: Point =>
-          val k = mapTransform(p)
-          clipToKey(k, pointPredicates).toSeq.iterator
-        case mp: MultiPoint =>
-          mp.points
-            .map(mapTransform(_))
-            .distinct
-            .map(clipToKey(_, mpOrLinePredicates))
-            .flatten
-            .iterator
-        case l: Line =>
-          mapTransform.multiLineToKeys(MultiLine(l))
-            .map(clipToKey(_, mpOrLinePredicates))
-            .flatten
-        case ml: MultiLine =>
-          mapTransform.multiLineToKeys(ml)
-            .map(clipToKey(_, mpOrLinePredicates))
-            .flatten
-        case p: Polygon =>
+        /* Points will always exist inside their Extent */
+        case p:  Point      => keys.iterator.map(k => (k, feature))
+        case mp: MultiPoint => keys.iterator.flatMap(k => clipToKey(k, mpOrLinePredicates))
+        case l:  Line       => keys.iterator.flatMap(k => clipToKey(k, mpOrLinePredicates))
+        case ml: MultiLine  => keys.iterator.flatMap(k => clipToKey(k, mpOrLinePredicates))
+        case p:  Polygon if keys.size > 10 =>
```

Comment:
Pulled 10 out of thin air. Actually now that I think about it, if `keys.size > 8`, then the Polygon will always cover the Extent and we never have to prepare the geom nor do intersections. We only need to *actually* check for covers when the key size is between 8 and 2... in which case would we even both with preparing it?

Reply:
Not true. What if the Polygon stretched outside its Extent (and so had many keys), but had a hole inside the Extent? In that case, returning just the `Extent` as the faux-clipped Geom would be a mistake.

Reply:
Most Polygons in OSM don't have holes, though. It feels like we can still use that information to save tons of operations.

Reply:
https://geotrellis.github.io/scaladocs/latest/index.html#geotrellis.vector.Polygon@hasHoles:Boolean

If no, then the `keys.size > 8` should always be reliable.

Reply:
https://github.com/locationtech/geotrellis/blob/master/vector/src/main/scala/geotrellis/vector/Polygon.scala#L107

I wonder if this is `O(1)`.

Reply:
Doesn't matter, actually. The `covers` check still needs to be performed *per Extent*, since each Extent varies in shape. The number of such Extents doesn't matter.

Reply:
Yeah, can't use the number of keys for `covers` because geometry could always be skinnier than the extent no matter the key count.